neqo-transport/src/packet/mod.rs

```diff
@@ -118,26 +118,32 @@ struct PacketBuilderOffsets {
 
 /// A packet builder that can be used to produce short packets and long packets.
 /// This does not produce Retry or Version Negotiation.
-pub struct PacketBuilder {
-    encoder: Encoder,
+pub struct PacketBuilder<'a> {
+    encoder: Encoder<'a>,
     pn: PacketNumber,
     header: Range<usize>,
     offsets: PacketBuilderOffsets,
+    /// The size limit of [`Self::encoder`], i.e. the maximum number of bytes to
+    /// be encoded into [`Self::encoder`]. Note that [`Self::encoder`] might
+    /// contain one or more packets already.
     limit: usize,
```

Comment:
Previously we would allocate a new `Vec` for each UDP datagram, with the capacity of the `Vec` set to the limit.

Now we have one long-lived receive buffer, with a fixed capacity. In order to enforce the limit `PacketBuilder` now owns the limit in its `limit: usize`, i.e. is explicit about it, instead of implicitly depending on the capacity of the `Vec` of `PacketBuilder::encoder`.